racket/collects/racket/treelist.rkt

```diff
@@ -0,0 +1,1334 @@
+       [(fx< (node-size n) MAX_WIDTH)
+        (Node (vector*-add-right (node-children n)
+                           (new-branch el (fx- height 1)))
+              (let ([sizes (node-sizes n)])
+                (and sizes
+                     (vector*-add-right sizes
+                                        (fx+ (vector*-ref sizes (fx- (vector*-length sizes) 1)) 1)))))]
+       [else
+        #false])]))
+
+;; create a branch of height `height` terminating in a unary leaf node containing `el`
+(define (new-branch el height)
+  (if (fx= height 0)
+      (leaf el)
+      (Node (vector (new-branch el (fx- height 1))))))
+
+(define (treelist-map tl proc)
+  (check-treelist 'treelist-map tl)
+  (unless (and (procedure? proc) (procedure-arity-includes? proc 1))
+    (raise-argument-error 'treelist-map "(procedure-arity-includes/c 1)" proc))
+  (for/treelist ([v (in-treelist tl)])
+    (proc v)))
+
+(define (treelist-for-each tl proc)
+  (check-treelist 'treelist-for-each tl)
+  (unless (and (procedure? proc) (procedure-arity-includes? proc 1))
+    (raise-argument-error 'treelist-for-each "(procedure-arity-includes/c 1)" proc))
+  (for ([v (in-treelist tl)])
+    (proc v)))
+
+(define (treelist-member? tl v [eql? equal?])
+  (check-treelist 'treelist-member? tl)
+  (unless (and (procedure? eql?) (procedure-arity-includes? eql? 2))
+    (raise-argument-error 'treelist-member? "(procedure-arity-includes/c 2)" eql?))
+  (for/fold ([r #f]) ([el (in-treelist tl)]
+                      #:when (eql? v el)
+                      #:final #t)
+    #t))
+
+(define (treelist-find tl match?)
+  (check-treelist 'treelist-find tl)
+  (unless (and (procedure? match?) (procedure-arity-includes? match? 1))
+    (raise-argument-error 'treelist-find "(procedure-arity-includes/c 1)" match?))
+  (for/fold ([r #f]) ([el (in-treelist tl)]
+                      #:when (match? el)
+                      #:final #t)
+    el))
+
+(define (check-sort-arguments who less-than? get-key)
+  (unless (and (procedure? less-than?) (procedure-arity-includes? less-than? 2))
+    (raise-argument-error who "(procedure-arity-includes/c 2)" less-than?))
+  (unless (or (not get-key)
+              (and (procedure? get-key) (procedure-arity-includes? get-key 1)))
+    (raise-argument-error who "(procedure-arity-includes/c 1)" get-key)))
+
+(define (treelist-sort tl less-than?
+                       #:key [get-key #f]
+                       #:cache-keys? [cache-keys? #f])
+  (check-treelist 'treelist-sort tl)
+  (check-sort-arguments 'treelist-sort less-than? get-key)
+  (define vec (treelist->vector tl))
+  (vector-sort! vec less-than? 0 (vector*-length vec) get-key cache-keys?)
+  (vector->treelist vec))
+
+(define-values (prop:treelist-chaperone
+                treelist-chaperone?
+                treelist-chaperone-ref)
+  (make-impersonator-property 'treelist))
+
+(struct treelist-wrapper (chaperone? prev ref set insert append))
```

Comment:
Should this be `#:sealed` and/or `#:authentic`?

Reply:
That seems useful to generate less code, so I added those.